debian/honeyBadger.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.47.4.
.TH FLAG "1" "September 2016" "flag provided but not defined: -version" "User Commands"
.SH NAME
flag \- manual page for flag provided but not defined: -version
.SH DESCRIPTION
.SS "Usage of ./honeyBadger:"
.HP
\fB\-archive_dir\fR string
.IP
archive directory for storing attack logs and related pcap files
.HP
\fB\-connection_max_buffer\fR int
.PP
        
Max packets to buffer for a single connection before skipping over a gap in data
and continuing to stream the connection after the buffer.  If zero or less, this
is infinite. (default 100)
.HP
\fB\-daq\fR string
.IP
Data AcQuisition packet source: pcapgo, libpcap, AF_PACKET or BSD_BPF.
.PP
BSD_BPF is BSD systems only.
AF_PACKET is Linux only.
libpcap builds on Linux and FreeBSD and can read every kind of pcap format.
pcapgo builds on every platform but does not support pcap\-ng format, only pcap v2.4.
.IP
(default "libpcap")
.HP
\fB\-detect_coalesce_injection\fR
.IP
Detect coalesce injection attacks (default true)
.HP
\fB\-detect_hijack\fR
.IP
Detect handshake hijack attacks (default true)
.HP
\fB\-detect_injection\fR
.IP
Detect injection attacks (default true)
.HP
\fB\-f\fR string
.IP
BPF filter for pcap (default "tcp")
.HP
\fB\-i\fR string
.IP
Interface to get packets from (default "eth0")
.HP
\fB\-l\fR string
.IP
incoming log dir used initially for pcap files if packet logging is enabled
.HP
\fB\-log_packets\fR
.IP
if set to true then log all packets for each tracked TCP connection
.HP
\fB\-max_concurrent_connections\fR int
.IP
Maximum number of concurrent connection to track. (default 300)
.HP
\fB\-max_pcap_log_size\fR int
.IP
maximum pcap size per rotation in megabytes (default 10)
.HP
\fB\-max_pcap_rotations\fR int
.IP
maximum number of pcap rotations per connection (default 100)
.HP
\fB\-max_ring_packets\fR int
.IP
Max packets per connection stream ring buffer (default 40)
.HP
\fB\-metadata_attack_log\fR
.IP
if set to true then attack reports will only include metadata
.HP
\fB\-pcapfile\fR string
.IP
pcap filename to read packets from rather than a wire interface.
.PP
This option is to be combined with a \fB\-daq=\fR setting of either "pcapgo" OR "libpcap"!
.HP
\fB\-s\fR int
.IP
SnapLen for pcap packet capture (default 65536)
.HP
\fB\-tcp_idle_timeout\fR duration
.IP
tcp idle timeout duration (default 10m0s)
.HP
\fB\-total_max_buffer\fR int
.PP
        
Max packets to buffer total before skipping over gaps in connections and
continuing to stream connection data.  If zero or less, this is infinite (default 1000)
.HP
\fB\-w\fR string
.IP
timeout for reading packets off the wire (default "3s")
.SS "Usage of ./honeyBadger:"
.HP
\fB\-archive_dir\fR string
.IP
archive directory for storing attack logs and related pcap files
.HP
\fB\-connection_max_buffer\fR int
.PP
        
Max packets to buffer for a single connection before skipping over a gap in data
and continuing to stream the connection after the buffer.  If zero or less, this
is infinite. (default 100)
.HP
\fB\-daq\fR string
.IP
Data AcQuisition packet source: pcapgo, libpcap, AF_PACKET or BSD_BPF.
.PP
BSD_BPF is BSD systems only.
AF_PACKET is Linux only.
libpcap builds on Linux and FreeBSD and can read every kind of pcap format.
pcapgo builds on every platform but does not support pcap\-ng format, only pcap v2.4.
.IP
(default "libpcap")
.HP
\fB\-detect_coalesce_injection\fR
.IP
Detect coalesce injection attacks (default true)
.HP
\fB\-detect_hijack\fR
.IP
Detect handshake hijack attacks (default true)
.HP
\fB\-detect_injection\fR
.IP
Detect injection attacks (default true)
.HP
\fB\-f\fR string
.IP
BPF filter for pcap (default "tcp")
.HP
\fB\-i\fR string
.IP
Interface to get packets from (default "eth0")
.HP
\fB\-l\fR string
.IP
incoming log dir used initially for pcap files if packet logging is enabled
.HP
\fB\-log_packets\fR
.IP
if set to true then log all packets for each tracked TCP connection
.HP
\fB\-max_concurrent_connections\fR int
.IP
Maximum number of concurrent connection to track. (default 300)
.HP
\fB\-max_pcap_log_size\fR int
.IP
maximum pcap size per rotation in megabytes (default 10)
.HP
\fB\-max_pcap_rotations\fR int
.IP
maximum number of pcap rotations per connection (default 100)
.HP
\fB\-max_ring_packets\fR int
.IP
Max packets per connection stream ring buffer (default 40)
.HP
\fB\-metadata_attack_log\fR
.IP
if set to true then attack reports will only include metadata
.HP
\fB\-pcapfile\fR string
.IP
pcap filename to read packets from rather than a wire interface.
.PP
This option is to be combined with a \fB\-daq=\fR setting of either "pcapgo" OR "libpcap"!
.HP
\fB\-s\fR int
.IP
SnapLen for pcap packet capture (default 65536)
.HP
\fB\-tcp_idle_timeout\fR duration
.IP
tcp idle timeout duration (default 10m0s)
.HP
\fB\-total_max_buffer\fR int
.PP
        
Max packets to buffer total before skipping over gaps in connections and
continuing to stream connection data.  If zero or less, this is infinite (default 1000)
.HP
\fB\-w\fR string
.IP
timeout for reading packets off the wire (default "3s")
.SH "SEE ALSO"
The full documentation for
.B flag
is maintained as a Texinfo manual.  If the
.B info
and
.B flag
programs are properly installed at your site, the command
.IP
.B info flag
.PP
should give you access to the complete manual.


debian/changelog
honeybadger (0.0~git20160224.27c83ff-1) UNRELEASED; urgency=medium

  * Initial release. (Closes: #XXXXXX)

 -- David Stainton <dstainton415@gmail.com>  Mon, 01 Feb 2016 17:02:40 +0100


tools/honeyBadger.go
/*
 *    honeyBadger.go - HoneyBadger core program for detecting TCP attacks
 *    such as handshake-hijack, segment veto and sloppy injection.
 *
 *    Copyright (C) 2014  David Stainton
 *
 *    This program is free software: you can redistribute it and/or modify
 *    it under the terms of the GNU General Public License as published by
 *    the Free Software Foundation, either version 3 of the License, or
 *    (at your option) any later version.
 *
 *    This program is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU General Public License for more details.
 *
 *    You should have received a copy of the GNU General Public License
 *    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package main

import (
	"flag"
	"fmt"
	"github.com/david415/HoneyBadger"
	"log"
	"os"
	"os/signal"
	"path/filepath"
)

func main() {
	var (
		iface       = flag.String("i", "eth0", "Interface to get packets from")
		snaplen     = flag.Int("s", 65536, "SnapLen for pcap packet capture")
		filter      = flag.String("f", "tcp", "BPF filter for pcap")
		logDir      = flag.String("l", "honeyBadger-logs", "log directory")
		isDaemonLog = flag.Bool("d", false, "daemon-log")
	)
	flag.Parse()

	if *isDaemonLog {
		f, err := os.OpenFile(filepath.Join(*logDir, "honeyBadger.log"), os.O_RDWR|os.O_CREATE|os.O_APPEND, 0666)
		if err != nil {
			panic(fmt.Sprintf("error opening file: %v", err))
		}
		defer f.Close()
		log.SetOutput(f)
	}

	service := HoneyBadger.NewHoneyBadgerService(*iface, *filter, *snaplen, *logDir)

	log.Println("HoneyBadger: comprehensive TCP injection attack detection.")

	// quit when we detect a control-c
	c := make(chan os.Signal, 1)
	signal.Notify(c, os.Interrupt)
	<-c
	service.Stop()
}
